src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -20,6 +20,169 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
+                        IList<TSource> list => new ListPartition<TSource>(list, startIndex, endIndex - 1),
+                        _ => new EnumerablePartition<TSource>(source, startIndex, endIndex - 1)
+                    };
+            }
+
+            return TakeIterator(source, isStartIndexFromEnd, startIndex, isEndIndexFromEnd, endIndex);
+        }
+
+        private static IEnumerable<TSource> TakeIterator<TSource>(
+            IEnumerable<TSource> source, bool isStartIndexFromEnd, int startIndex, bool isEndIndexFromEnd, int endIndex)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(startIndex >= 0);
+            Debug.Assert(endIndex >= 0);
+
+            if (isStartIndexFromEnd)
+            {
+                if (startIndex == 0)
+                {
+                    yield break;
+                }
+
+                using IEnumerator<TSource> e = source.GetEnumerator();
+                if (!e.MoveNext())
+                {
+                    yield break;
+                }
+
+                int index = 0;
+                Queue<TSource> queue = new();
+                queue.Enqueue(e.Current);
+
+                while (e.MoveNext())
+                {
+                    checked
+                    {
+                        index++;
+                    }
+
+                    if (queue.Count == startIndex)
+                    {
+                        queue.Dequeue();
+                    }
+
+                    queue.Enqueue(e.Current);
+                }
+
+                int count = checked(index + 1);
+                Debug.Assert(queue.Count == Math.Min(count, startIndex));
+
+                startIndex = count - startIndex;
+                if (startIndex < 0)
+                {
+                    startIndex = 0;
+                }
+
+                if (isEndIndexFromEnd)
+                {
+                    endIndex = count - endIndex;
+                }
+                else if (endIndex > count)
+                {
+                    endIndex = count;
+                }
+
+                Debug.Assert(endIndex - startIndex <= queue.Count);
+                for (int rangeIndex = startIndex; rangeIndex < endIndex; rangeIndex++)
+                {
+                    yield return queue.Dequeue();
+                }
+            }
+            else
+            {
+                using IEnumerator<TSource> e = source.GetEnumerator();
+                if (!e.MoveNext())
+                {
+                    yield break;
+                }
+
+                int index = 0;
+                while (index < startIndex && e.MoveNext())
+                {
+                    checked
+                    {
+                        index++;
+                    }
+                }
+
+                if (index != startIndex)
+                {
+                    yield break;
+                }
```

Comment:
Can't this just be:
```C#
int index = 0;
while (index < startIndex)
{
    if (e.MoveNext())
    {
       yield break; // or "return Empty<TSource>();" per my previous comment about making this not an iterator
    }

    checked { index++; }
}
```
?

Reply:
I feel my version is shorter and easier to read. Anyway, I updated the code since requested.